### DISCUSSION TITLE
Why does the decorator pattern not include ACK messages? Are there any other reasons for this?

### DISCUSSION THREAD

**Discussion by ChuLiqiang** [resiprocate/resiprocate#355] [Q&A] [answered]
![image](https://github.com/resiprocate/resiprocate/assets/2747722/5edc2815-ef6a-4d67-92cc-ae143d325ad5)

**Comment by sgodin** [comment, marked as answer]
It does include Ack's.

Failure ACKs are part of the same transaction, so copying them over makes sense.  Success ACK's (ACK/200) don't use the the same transaction-id as the request, and have are considered to be a unique transaction.  

DUM will add the outbound decorator you define in the Profile/UserProfile to all messages that go through DialogUsageManager::send and have an associated Dialogset with assigned UserProfile.  Otherwise it falls back to the outbound decorators defined in the MasterProfile.  This includes ACK/200's.

Scott

**Comment by ChuLiqiang** [comment]
> It does include Ack's.
> 
> Failure ACKs are part of the same transaction, so copying them over makes sense. Success ACK's (ACK/200) don't use the the same transaction-id as the request, and have are considered to be a unique transaction.
> 
> DUM will add the outbound decorator you define in the Profile/UserProfile to all messages that go through DialogUsageManager::send and have an associated Dialogset with assigned UserProfile. Otherwise it falls back to the outbound decorators defined in the MasterProfile. This includes ACK/200's.
> 
> Scott

I'm very sorry, I didn't quite understand what you meant. Actually, I just wanted to know if modifying it this way would cause any bugs. I know you've probably already answered my question. But since my understanding is not as deep as yours, I actually didn't understand the meaning, I'm very sorry. I was just very puzzled why the incorrect ACK message is not sent to the decorator for decoration. I really didn't understand much about userprofile and masterprofile.

I konw,you mean 200 ok ack can include in outbound  decorator. fail ack not include in outbound  decorator.I try remove Failure ACKs check in code.I find all ting is ok. Can you tell me if have some bug when i remote this check？

**Reply by sgodin** [reply]
You don't need to remove this check you just need to set copyToStackFailureAcks() on your decorator to true.

**Reply by ChuLiqiang** [reply]
thank you very much.
